module/VuFindSearch/src/VuFindSearch/Backend/Blender/Response/Json/RecordCollection.php

```diff
@@ -0,0 +1,344 @@
+ * @category VuFind
+ * @package  Search
+ * @author   David Maus <maus@hab.de>
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     http://vufind.org
+ */
+class RecordCollection
+    extends \VuFindSearch\Backend\Solr\Response\Json\RecordCollection
+{
+    /**
+     * Blender configuration
+     *
+     * @var \Laminas\Config\Config
+     */
+    protected $config;
+
+    /**
+     * Mappings configuration
+     *
+     * @var array
+     */
+    protected $mappings;
+
+    /**
+     * Any errors encountered
+     *
+     * @var array
+     */
+    protected $errors = [];
+
+    /**
+     * Constructor
+     *
+     * @param \Laminas\Config\Config $config   Configuration
+     * @param array                  $mappings Mappings configuration
+     */
+    public function __construct($config = null, $mappings = [])
+    {
+        $this->config = $config;
+        $this->mappings = $mappings;
+        $this->response = static::$template;
+    }
+
+    /**
+     * Initialize blended results
+     *
+     * @param RecordCollectionInterface $primaryCollection   Primary record
+     * collection
+     * @param RecordCollectionInterface $secondaryCollection Secondary record
+     * collection
+     * @param int                       $offset              Results list offset
+     * @param int                       $limit               Result limit
+     * @param int                       $blockSize           Record block size
+     *
+     * @return void
+     */
+    public function initBlended(
+        RecordCollectionInterface $primaryCollection = null,
+        RecordCollectionInterface $secondaryCollection = null,
+        $offset,
+        $limit,
+        $blockSize
+    ) {
+        $this->response = static::$template;
+        $this->response['response']['numFound']
+            = ($primaryCollection ? $primaryCollection->getTotal() : 0)
+            + ($secondaryCollection ? $secondaryCollection->getTotal() : 0);
+        $this->offset = $this->response['response']['start'] = $offset;
+        $this->rewind();
+
+        $primaryRecords = $primaryCollection ? $primaryCollection->getRecords() : [];
+        $secondaryRecords = $secondaryCollection
+            ? $secondaryCollection->getRecords() : [];
+
+        $label = $this->config['Primary']['label'] ?? '';
+        foreach ($primaryRecords as &$record) {
+            $record->setExtraDetail('blendSource', 'primary');
+            $record->setSourceIdentifiers(
+                $record->getSourceIdentifier(),
+                $primaryCollection->getSourceIdentifier()
+            );
+            if ($label) {
+                $record->addLabel($label, 'source');
+            }
+        }
+        unset($record);
+
+        $label = $this->config['Secondary']['label'] ?? '';
+        foreach ($secondaryRecords as &$record) {
+            $record->setExtraDetail('blendSource', 'secondary');
```

Comment:
This record processing is pretty repetitive; maybe you can create a support method to simplify this.